matcher.go

```diff
@@ -0,0 +1,185 @@
+package elemental
+
+import (
+	"fmt"
+	"reflect"
+)
+
+// MatchesFilter determines whether an identity matches a filter
+func MatchesFilter(identifiable AttributeSpecifiable, filter *Filter, opts ...MatcherOption) (bool, error) {
+
+	if filter == nil {
+		panic(fmt.Errorf("elemental: filter cannot be nil"))
+	}
+
+	if identifiable == nil {
+		panic(fmt.Errorf("elemental: identifiable cannot be nil"))
+	}
+
+	matched := true
+	var err error
+
+	for i, op := range filter.Operators() {
+		switch op {
+		case AndOperator:
+			comparator := filter.Comparators()[i]
+			attributeName := filter.Keys()[i]
+			attributeValue := identifiable.ValueForAttribute(attributeName)
+
+			// important: when adding support for a new comparator, you must only short circuit and return under two conditions:
+			//    - an error occurred by your comparator
+			//    - the comparator failed to find a match
+			//
+			// this is because we are dealing with AND semantics here; success is only guaranteed if all AndOperator's find a match
+
+			switch comparator {
+			case EqualComparator:
+				if equals, err := equals(attributeName, attributeValue, filter.Values()[i][0]); err != nil || !equals {
+					return false, err
+				}
+			case NotEqualComparator,
+				GreaterComparator,
+				GreaterOrEqualComparator,
+				LesserComparator,
+				LesserOrEqualComparator,
+				InComparator,
+				NotInComparator,
+				ContainComparator,
+				NotContainComparator,
+				MatchComparator,
+				NotMatchComparator,
+				ExistsComparator,
+				NotExistsComparator:
+				return false, fmt.Errorf("elemental: unsuported comparator %q", translateComparator(filter.Comparators()[i]))
+			default:
+				panic(fmt.Errorf("elemental: unknown comparator %q", translateComparator(filter.Comparators()[i])))
```

Comment:
FYI @primalmotion I can't write a unit test for this unless I create an interface for filter which I can mock.